meshmode/array_context.py

```diff
@@ -368,6 +387,41 @@ def tag_axis(self, iaxis, tags: Union[Sequence[Tag], Tag], array):
         .. versionadded:: 2021.2
         """
 
+    @memoize_method
+    def _get_einsum_prg(self, spec, arg_names, tagged):
+        return lp.make_einsum(
+            spec,
+            arg_names,
+            options=_DEFAULT_LOOPY_OPTIONS,
+            tags=tagged,
+        )
+
+    def einsum(self, spec, *args, arg_names=None, tagged=()):
+        """Executes a :mod:`loopy` einsum program on the arguments *args*.
+
+        :arg spec: a string denoting the subscripts for
+            summation as a comma-separated list of subscript labels.
+            This follows the usual :func:`numpy.einsum` convention.
+            Note that the explicit indicator `->` for the precise output
+            form is required.
+        :arg args: a sequence of array-like operands, whose order matches
+            the subscript labels provided by *spec*.
+        :arg arg_names: an optional iterable of string types denoting
+            the names of the *args*. If *None*, default names will be
+            generated.
+        :arg tagged: an optional iterable of :class:`pytools.tag.Tag`
+            objects for specifying kernel transformations.
```

Comment:
```suggestion
        :arg tagged: an optional sequence of :class:`pytools.tag.Tag`
            objects specifying the tags to be applied to the operation.
```
(It's not this routine's business to determine what the tags are used for.)